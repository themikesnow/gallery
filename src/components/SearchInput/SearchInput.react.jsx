import React from 'react';
import classnames from 'classnames';

import Loading from '../Loading/Loading.react';

export default class SearchInput extends React.Component {

  static propTypes = {
    isBusy: React.PropTypes.bool,
    onSearch: React.PropTypes.func,
  };

  static defaultProps = {
    isBusy: false,
    onSearch: () => {},
  };

  constructor(props) {
    super(props);
    this.onChange = ::this.onChange;
    this.onClick = ::this.onClick;
    this.onKeyPressed = ::this.onKeyPressed;
    this.onSearch = ::this.onSearch;
  }

  state = {
    value: '',
  };

  onChange(e) {
    this.setState({
      value: e.target.value,
    });
  }

  onClick() {
    this.onSearch();
  }

  onKeyPressed(e) {
    if (e.key === 'Enter') {
      this.onSearch();
    }
  }

  onSearch() {
    if (!this.props.isBusy && typeof this.props.onSearch === 'function') {
      this.props.onSearch(this.state.value);
    }
  }

  render() {
    return (
      <div className="search-input">
        <div className="controls">
          <input
            className="search-input-control"
            type="text"
            value={this.state.value}
            placeholder="Enter keyword..."
            onChange={this.onChange}
            onKeyPress={this.onKeyPressed}
          />
          <button className={classnames({ 'btn btn-submit': true, disabled: this.props.isBusy || !this.state.value })} onClick={this.state.value && this.onClick}>
            {this.props.isBusy ? <Loading isIcon />
              : <span>Search</span>
            }
          </button>
        </div>
      </div>
    );
  }
}
