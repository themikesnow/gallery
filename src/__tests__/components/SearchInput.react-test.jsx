import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import { expect as expectChai } from 'chai';
// import sinon from 'sinon';

import SearchInput from '../../components/SearchInput/SearchInput.react';

describe('<SearchInput />', () => {
  it('Match Snapshot', () => {
    const myMock = jest.fn();
    let component = renderer.create(<SearchInput onSearch={myMock} />);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();

    component = renderer.create(<SearchInput onSearch={myMock} isBusy />);
    tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should handle null onSearch prop when submit button clicked', () => {
    const wrapper = mount(<SearchInput onSearch={null} />);
    wrapper.find('.btn-submit').simulate('click');
    // No need to check function/mock...
  });

  it('should call onSearch when submit button clicked', () => {
    const myMock = jest.fn();
    const wrapper = mount(<SearchInput onSearch={myMock} />);
    wrapper.find('.btn-submit').simulate('click');
    expectChai(myMock.mock.calls.length).to.equal(1);
  });

  it('should call onChange', () => {
    const theValue = 'test';
    const myMock = jest.fn();
    const wrapper = mount(<SearchInput onSearch={myMock} />);

    wrapper.find('.search-input-control').simulate('change', { target: { value: theValue } });
    expectChai(wrapper.state().value).to.equal(theValue);
  });

  it('should call onSearch when enter pressed on the search input', () => {
    const myMock = jest.fn();
    const wrapper = mount(<SearchInput onSearch={myMock} />);
    wrapper.find('.search-input input').simulate('keyPress', { key: 'Enter' });
    expectChai(myMock.mock.calls.length).to.equal(1);
  });

  it('should Not call onSearch when key (not enter) pressed on the search input', () => {
    const myMock = jest.fn();
    const wrapper = mount(<SearchInput onSearch={myMock} />);
    wrapper.find('.search-input input').simulate('keyPress', { key: 'Backspace' });
    expectChai(myMock.mock.calls.length).to.equal(0);
  });
});
