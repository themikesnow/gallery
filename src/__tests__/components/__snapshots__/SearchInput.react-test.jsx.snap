// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SearchInput /> Match Snapshot 1`] = `
<div
  className="search-input"
>
  <div
    className="controls"
  >
    <input
      className="search-input-control"
      onChange={[Function]}
      onKeyPress={[Function]}
      placeholder="Enter keyword..."
      type="text"
      value=""
    />
    <button
      className="btn btn-submit disabled"
      onClick={[Function]}
    >
      <span>
        Search
      </span>
    </button>
  </div>
</div>
`;

exports[`<SearchInput /> Match Snapshot 2`] = `
<div
  className="search-input"
>
  <div
    className="controls"
  >
    <input
      className="search-input-control"
      onChange={[Function]}
      onKeyPress={[Function]}
      placeholder="Enter keyword..."
      type="text"
      value=""
    />
    <button
      className="btn btn-submit disabled"
      onClick={[Function]}
    >
      <i
        className="fa fa-refresh fa-spin fa-lg fa-fw"
      />
    </button>
  </div>
</div>
`;
